common-lib/src/cmr/common/generics.clj

```diff
@@ -117,3 +119,61 @@
           {}
           (latest-approved-documents)))
+
+;HTML templates for generic table of content items
+(def ingest-table-of-contents-template (slurp (io/resource "ingest-table-of-contents-template.txt")))
+(def search-table-of-contents-template (slurp (io/resource "search-table-of-contents-template.txt")))
+
+(defn read-generic-doc-file
+  "Return the specific schema's documentation files given the schema keyword name and version number.
+   if the file cannot be read, return an empty string which will have no impact on the API document.
+   Parameters:
+   * file-name: [ingest | search]
+   * generic-keyword: [:grid | ...]
+   * generic-version: 0.0.1
+   Returns: string"
+  [file-name generic-keyword generic-version]
+  (try
+    (-> "schemas/%s/v%s/%s.md"
+        (format (name generic-keyword) generic-version (name file-name))
+        (io/resource)
+        (slurp))
+    (catch Exception e (str ""))))
+
+(defn all-generic-docs
+  "Parse over all of the generic documents and return their combined markdown as a string
+  Parameters:
+  * file-name: [ingest | search]
+  Returns: string"
+  [file-name]
+  (string/join (seq (for [[k,v] (latest-approved-documents)] (read-generic-doc-file file-name k (str v))))))
+
+(defn get-generics-with-documenation
+  "Retrieve the names of the generics that have documentation files. Re-seq returns both the full match and the group, so parse out all of the odd numbered indicies
+   so that only the group matches remain
+   Parameters:
+   * generic-markdown: a slurped markdown file
+   Returns: string"
+  [generic-markdown]
+  (take-nth 2 (rest (flatten (re-seq #"### <a name=\"create-update-([a-z]+)" generic-markdown)))))
+
+(defn table-of-contents-html
+  "Return the html for the table of contents given a generic concept
+   Parameters:
+   * table-of-contents-template: [ingest-table-of-contents-template | search-table-of-contents-template]
+   Returns: string"
+  [table-of-contents-template generic-type]
+  (-> table-of-contents-template
+      (string/replace #"%generic%" generic-type)
+      (string/replace  #"%uc-generic%" (string/capitalize generic-type))
+      (string/replace  #"%plural-generic%" (inf/plural generic-type))
+      (string/replace  #"%uc-plural-generic%" (inf/plural (string/capitalize generic-type)))))
+
+(defn all-generic-table-of-contents
+  "Parse over all of the generic documents and return their combined html for the table of contents as a string
+   we use ingest to retrieve the list of generics with documenation for both search and ingest as it would be expected both documents would be in the schema dir
+   Parameters:
+   * table-of-contents-template: [ingest-table-of-contents-template | search-table-of-contents-template]
+   Returns: string"
+  [table-of-contents-template]
+  (string/join (mapv #(table-of-contents-html table-of-contents-template %) (get-generics-with-documenation (all-generic-docs "ingest")))))
```

Comment:
so why are you not looking for ## tags in the markdown and turning those into table of content links?